Prevent uncontrolled crash in Release mode when ParentDataWidget is used incorrectly

darajava opened the pull request:
Fixes #108186 and #155051

The following (incorrect) code causes a nasty crash in release mode:

```dart
void main() =>
  runApp(
    const Flexible( // <- This is incorrect, should be child of `Row` or similar
      child:
        Text('This is only shown in debug mode!',
          key: Key('title'),
          textDirection: TextDirection.ltr,
        ),
      ),
    );
```

Before:

![image](https://github.com/user-attachments/assets/f9750e93-f716-418b-b6dc-9f3644df1afe)

After:

<img width="739" alt="image" src="https://github.com/user-attachments/assets/e9377d6a-2405-4878-9e79-34955f5ea0dd">

I have refactored `_updateParentData` to make it more simple and to give a clearer and more specific error message when `Flexible` or `Expanded` is misused. In both release and debug builds we now check `parentDataWidget.debugIsValidRenderObject` before calling `applyParentData`. The problem in the previous code was that there was a check happening in an `assert`. Therefore, in release/profile modes `applyParentData` was being unconditionally called, leading to an error causing the UI to completely break and become unusable.

With this change, `Flexible` can be called now in any component without breaking the UI, and still gets an exception reported in debug mode. 

I had initially tried to show `ErrorWidget` when the tree is in a broken state, but that leads to more exceptions as the original comment in the code I worked on suggested because adding more widgets to a broken tree leads to more exceptions which is hard to test and poor for the debugging experience, as is mentioned. 

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

darajava committed:
Don't ignore error on debug mode when ParentDataWidget is used incorrectly

darajava commented:
This is my attempt at testing:

```
  testWidgets('Incorrect use of ParentDataWidget shows a useful error message', (WidgetTester tester) async {
    await tester.pumpWidget(
      Container(
        key: const Key('container'),
        child: Flexible(
          child: Container(),
        ) // Flexible should not be used directly in a Container or other inappropriate parents
      ),
    );

    // await tester.pumpAndSettle();

    final dynamic exception = tester.takeException();


    // Expect a FlutterError to be thrown
    expect(exception, isFlutterError);
  });
```

I "pump" an incorrect widget, expecting an exception, but I actually get 3 exceptions which all seem to be `FlutterError`s. However, I get the following output from the above test:

```
Expected: <Instance of 'FlutterError'>
  Actual: 'Multiple exceptions (3) were detected during the running of the current test, and at
least one was unexpected.'
   Which: is not an instance of 'FlutterError'
```

Why is this code giving 3 exceptions? I'm pretty sure as well that `at least one was unexpected` is wrong, since they are all `FlutterError`s.

Could I get some help here?

darajava committed:
put back into original state to check if CI tests pass

darajava committed:
Merge branch 'master' into improve-debug-release-disparity

darajava committed:
re-add fix

darajava committed:
fix a different way

darajava committed:
Refactor _updateParentData and give clearer error.

darajava committed:
Add unit test to verify exception thrown in debug mode.

darajava committed:
remove debug

darajava committed:
re-add test

darajava committed:
Merge branch 'master' into improve-debug-release-disparity

navaronbracke commented on packages/flutter/test/widgets/framework_test.dart:
```diff
@@ -1574,6 +1574,23 @@ void main() {
+
+    final dynamic exception = tester.takeException();
+
+    expect(exception, isFlutterError);
```
I would prevent the use of `dynamic` here. (If you really want a type, use `Object?` instead)

```suggestion
    expect(tester.takeException(), isFlutterError);
```

navaronbracke replied:
You should also check if the error message is correct, using the `having()` matcher.

navaronbracke commented on packages/flutter/test/widgets/framework_test.dart:
```diff
@@ -1574,6 +1574,23 @@ void main() {
 
+  testWidgets('Incorrect use of ParentDataWidget shows a useful error message', (WidgetTester tester) async {
+    await tester.pumpWidget(
+      Container(
```
Nit: Use `SizedBox` instead of `Container`.

navaronbracke commented on packages/flutter/lib/src/widgets/framework.dart:
```diff
@@ -6688,31 +6688,36 @@ abstract class RenderObjectElement extends Element {
+        try {
           throw FlutterError.fromParts(<DiagnosticsNode>[
-            ErrorSummary('Incorrect use of ParentDataWidget.'),
+            ErrorSummary('Incorrect use of ParentDataWidget. The widget `${
```
Could the formatting be improved by using a triple quotes string?

I.e.

```
'''
Incorrect use of ParentDataWidget. 
The widget `${f.runtimeType} must be a direct child
of a `Row`, `Column`, or `Flex` widget.'
'''
```

navaronbracke commented on packages/flutter/lib/src/widgets/framework.dart:
```diff
@@ -6688,31 +6688,36 @@ abstract class RenderObjectElement extends Element {
           throw FlutterError.fromParts(<DiagnosticsNode>[
-            ErrorSummary('Incorrect use of ParentDataWidget.'),
+            ErrorSummary('Incorrect use of ParentDataWidget. The widget `${
+              parentDataWidget.runtimeType
```
Calling `runtimeType.toString()` in release mode is expensive. Consider using the `objectRuntimeType()` helper, although that one also uses an assert.

See https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/foundation/object.dart

darajava replied:
I found a `toStringShort` helper function within this file which calls `objectRuntimeType`

darajava committed:
Resolve review comments
- don't call expensive `.runtimeType`
- improve formatting
- make test more specific

darajava committed:
Merge branch 'master' into improve-debug-release-disparity

darajava commented:
@navaronbracke thanks for your review, I have resolved your comments and updated the branch.

darajava committed:
Fix CI and improve formatting

darajava committed:
Merge branch 'master' into improve-debug-release-disparity

navaronbracke commented on packages/flutter/test/widgets/framework_test.dart:
```diff
@@ -1574,6 +1574,31 @@ void main() {
+      SizedBox(
+        key: const Key('container'),
+        child: Flexible( // Flexible should not be used directly in a Container or other inappropriate parents
+          child: Container(),
```
This is still a `Container`. This should probably also be a `SizedBox()`. See https://dart.dev/tools/linter-rules/avoid_unnecessary_containers

You might also need to add `const` to the first SizedBox widget.

navaronbracke commented on packages/flutter/test/widgets/framework_test.dart:
```diff
@@ -1574,6 +1574,31 @@ void main() {
+    await tester.pumpWidget(
+      SizedBox(
+        key: const Key('container'),
+        child: Flexible( // Flexible should not be used directly in a Container or other inappropriate parents
```
If you want to place a comment here, you can put it above the Flexible. That makes the formatting a bit nicer.
```suggestion
        // Flexible should only be used inside Flex widgets, such as a Column or Row.
        child: Flexible(
```

navaronbracke commented on packages/flutter/lib/src/widgets/framework.dart:
```diff
@@ -6688,31 +6688,41 @@ abstract class RenderObjectElement extends Element {
+        try {
+          String errorMessage = "Incorrect use of ParentDataWidget.\n";
+
+          if (parentDataWidget.toStringShort() == 'Positioned') {
```
Why do we special case the Positioned/Stack case here? There might be an unknown amount of possibilities that lead to the "Incorrect use of ParentDataWidget" error.

darajava replied:
I remember when I was new to Flutter, the `Incorrect use of ParentDataWidget.` was very confusing to me, so I wanted to have a more descriptive error message as well as to fix the CI tests.

I'll add specific cases only for `Positioned`, `Flexible`, and `Expanded`, leaving the generic error if it doesn't match.

darajava committed:
Move comment to better place

Co-authored-by: Navaron Bracke <brackenavaron@gmail.com>

darajava committed:
don't catch generic case for error message

darajava committed:
fix CI

darajava committed:
Use correct `toString` method

darajava committed:
fix wrong variable in interpolated string

darajava committed:
Merge branch 'master' into improve-debug-release-disparity

navaronbracke commented on packages/flutter/lib/src/widgets/framework.dart:
```diff
@@ -6688,31 +6688,42 @@ abstract class RenderObjectElement extends Element {
+          String errorMessage = 'Incorrect use of ParentDataWidget.\n';
+
+          if (parentDataWidgetType == 'Positioned') {
+            errorMessage += 'The widget `Positioned` must be a descendent of a `Stack` widget.\n';
```
The wording in these two messages is different? One mentions `descendant` (there is a typo in that word, too)
and the other mentions the word `child`.

darajava replied:
I'll update the spelling there. So with `Positioned` it can be anywhere in the `Stack` (i.e. it does not have to be a direct child). For a `Flexible` or `Expanded`, it needs to be a *direct* child, hence the differing wording.

navaronbracke commented on packages/flutter/test/widgets/framework_test.dart:
```diff
@@ -1574,6 +1574,32 @@ void main() {
     expect(RawKeyboard.instance.keyEventHandler, same(rawKeyEventHandler));
   });
 
+  testWidgets('Incorrect use of ParentDataWidget shows a useful error message', (WidgetTester tester) async {
```
Do we need an additional test for the Stack/Positioned case?

darajava replied:
So there are already tests for this which I've found in `parent_test_data.dart`, which already covers the case that I've tested in this PR. I've edited the tests in `parent_test_data` to work with the new error messages and I now think the test I have made is redundant, and actually would not have failed before I made the change in `framework.dart`, so I think I need a better test, what do you think?

I can't currently think of a way to test release-mode code though - which is what this PR addresses, so I'm going to look into this.

darajava replied:
The original code I modified relied on code in an `assert` to control the flow, and hence running into nasty exceptions only in release mode. I have made a contrived example below demonstrating the issue in the original code:

```dart
num returnDivisionCalculation() {
  num numerator = 1;
  num denominator = 0;
  bool attemptCalculation = true;

  assert(() {
    if (denominator == 0) {
      attemptCalculation = false;

      try {
        throw "Specific division error";
      } catch {
        // log the error and continue
      }
    }
  }());

  if (attemptCalculation) { // Always true in release mode.
    return numerator / denominator;
  }

  return 0;
}
```

In debug mode, the code in `assert` will be run, causing safe and controlled execution of the code, but in release mode an unexpected exception will be thrown.

I don't think it's possible to write a failing test that will succeed after my change and I will ask for an exemption on Discord. I have edited the tests to update the new error messages for misuse of these widgets (of which I am happy to revert, just thought I might improve them while here)

darajava committed:
Fix non-preferred spelling

darajava committed:
Merge branch 'master' into improve-debug-release-disparity

Piinks commented on packages/flutter/lib/src/widgets/framework.dart:
```diff
@@ -6688,31 +6688,42 @@ abstract class RenderObjectElement extends Element {
+          final String parentDataWidgetType = parentDataWidget.toStringShort();
+          String errorMessage = 'Incorrect use of ParentDataWidget.\n';
+
+          if (parentDataWidgetType == 'Positioned') {
+            errorMessage += 'The widget `Positioned` must be a descendant of a `Stack` widget.\n';
+          } else if (parentDataWidgetType == 'Flexible' || parentDataWidgetType == 'Expanded') {
+            errorMessage += 'The widget `$parentDataWidgetType` must be a direct child of a `Row`, `Column`, or `Flex` widget.\n';
+          }
+
```
This checking for types is not something we should do. There are many ParentDataWidgets, this error message handling could not possibly account for all of them. It also means we are baking in knowledge of other classes here.

darajava replied:
Sure, the idea here is to only account for 3 common ones -- it leaves the error message as-is for any other ParentDataWidget to reduce confusion. IMO "Incorrect use of ParentDataWidget" isn't a very clear error message.

Having said this I'm very happy to remove attempted improvements to DX and fix only the original issue!

darajava replied:
Ah I see, makes sense. I think that is beyond the scope of this fix then, would you agree? I will remove changes to the error messages in this PR tonight. 

A problem I'm having with this fix is testing. At its core, it's fixing some bad code that is only run in debug mode. I left a comment [here](https://github.com/flutter/flutter/pull/155091#discussion_r1766971811), do you know of a way to test this?

darajava committed:
Simplify PR scope
- Remove tests
- Remove attempt at improving error message

darajava commented:
@Piinks I've now simplified this PR, but it's still missing a test.

darajava committed:
Merge branch 'master' into improve-debug-release-disparity

chunhtai commented:
(triage) Hi @darajava we need a test to merge this pr. Can you add a test?

darajava commented:
Hi, no I don’t think this is testable since the bug is due to flow control
being determined from with an assert, which only is run on debug mode,
hence the bug only happens on release mode. Happy to add a test if there is
a way to do it!

On Wed, 2 Oct 2024 at 00:11, chunhtai ***@***.***> wrote:

> (triage) Hi @darajava <https://github.com/darajava> we need a test to
> merge this pr. Can you add a test?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/flutter/flutter/pull/155091#issuecomment-2387164666>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABNCOBN73HEEE5YSHHOS64LZZMMX5AVCNFSM6AAAAABODPODHOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDGOBXGE3DINRWGY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

chunhtai commented on packages/flutter/lib/src/widgets/framework.dart:
```diff
@@ -6688,11 +6688,16 @@ abstract class RenderObjectElement extends Element {
-      try {
-        if (!parentDataWidget.debugIsValidRenderObject(renderObject)) {
-          applyParentData = false;
+    if (parentDataWidget.debugIsValidRenderObject(renderObject)) {
```
This should probably not named started with debug, and should probably be convert into a private static method

dkwingsmt replied:
I agree. It affects the control flow, and should not be exclusive to assert mode. But it was written this way ever since the first day, and not only `debugIsValidRenderObject`, but also the method that it depends on, `debugTypicalAncestorWidgetClass`, are prefixed by `debug`, the latter being intended for overriding and is indeed overridden by many classes. We might want to ask Ian (the first author) whether it was intentional.

Hixie replied:
We shouldn't use debugTypicalAncestorWidgetClass in release mode, because it'll force widgets to be included that might otherwise get optimised out. But I think we can still throw in release mode even if the error message isn't as clear. We have lots of precedent for that.

Hixie replied:
(debugIsValidRenderObject should be renamed though)

darajava replied:
I've renamed `debugIsValidRenderObject` to make it static and private.

Are you saying that incorrect usage of ParentDataWidget should still throw and cause a crash to users in release mode? I'd have imagined that release mode should be as fault-tolerant as possible, and attempt to deliver a good experience even if it doesn't follow Flutter's rules. In my experience using e.g. `Flexible` outside a `Column` does not cause any visual/usability issues, only a broken tree which is flagged in the logs in debug mode. 

A similar case to this that current exists is using a `Positioned` outside a `Stack`. There is no point in using it outside a `Stack`, and it is an error, but it does not currently break the release build and gives an error in the console in debug mode.

darajava committed:
Merge branch 'master' into improve-debug-release-disparity

darajava committed:
rename method to check if render object is valid, and make private and static

darajava committed:
undo static

Piinks commented:
> Hi, no I don’t think this is testable since the bug is due to flow control
being determined from with an assert, which only is run on debug mode,
hence the bug only happens on release mode. Happy to add a test if there is
a way to do it!

@darajava in the test I would expect to verify the assertion is triggered.

Piinks commented on packages/flutter/lib/src/widgets/framework.dart:
```diff
@@ -1604,7 +1604,7 @@ abstract class ParentDataWidget<T extends ParentData> extends ProxyWidget {
   ///
   /// This is called just before [applyParentData] is invoked with the same
   /// [RenderObject] provided to that method.
-  bool debugIsValidRenderObject(RenderObject renderObject) {
```
This is a public method, making it private is a breaking change. Do we need to rename this here?

Piinks commented on packages/flutter/lib/src/widgets/framework.dart:
```diff
@@ -6688,11 +6688,16 @@ abstract class RenderObjectElement extends Element {
+      // If it's not a valid render object, then don't show the ErrorWidget, or
+      // throw an exception, just report the error and move on.
+      //
+      // This prevents it from crashing in release mode and still reports the
```
It may not prevent it from crashing, depending on what relies on that parent data being applied. For example, if a widget lays out multiple children, and relies on parent data to inform the layout, this change could skip a crash here, just to crash later when it cannot properly lay out the widgets.

darajava commented:
I’ve put too many hours into this important but simple fix and can’t afford to put more in unfortunately. If anyone wants to continue please feel free!
